bolt/lib/Target/AArch64/AArch64MCPlusBuilder.cpp

```diff
@@ -339,6 +369,183 @@ class AArch64MCPlusBuilder : public MCPlusBuilder {
+          It->getOperand(0).getReg() != AArch64::XZR ||
+          It->getOperand(3).getImm() != 0)
+        return std::nullopt;
+      TestedReg = It->getOperand(1).getReg();
+      ScratchReg = It->getOperand(2).getReg();
+
+      // Either XPAC(I|D) ScratchReg, ScratchReg
+      // or     XPACLRI
+      switch (StepAndGetOpcode()) {
+      default:
+        return std::nullopt;
+      case AArch64::XPACLRI:
+        // No operands to check, but using XPACLRI forces TestedReg to be X30.
+        if (TestedReg != AArch64::LR)
+          return std::nullopt;
+        break;
+      case AArch64::XPACI:
+      case AArch64::XPACD:
+        if (It->getOperand(0).getReg() != ScratchReg ||
+            It->getOperand(1).getReg() != ScratchReg)
+          return std::nullopt;
+        break;
+      }
+
+      // ORRXrs ScratchReg, XZR, TestedReg, 0 (used by "MOV reg, reg" alias)
+      if (StepAndGetOpcode() != AArch64::ORRXrs)
+        return std::nullopt;
+      if (It->getOperand(0).getReg() != ScratchReg ||
+          It->getOperand(1).getReg() != AArch64::XZR ||
+          It->getOperand(2).getReg() != TestedReg ||
+          It->getOperand(3).getImm() != 0)
+        return std::nullopt;
+
+      return std::make_pair(TestedReg, &*It);
+
+    case AArch64::TBZX:
+      // TBZX ScratchReg, 62, .Lon_success
+      ScratchReg = It->getOperand(0).getReg();
+      if (It->getOperand(1).getImm() != 62)
+        return std::nullopt;
+      // Not checking .Lon_success (see above).
+
+      // EORXrs ScratchReg, TestedReg, TestedReg, 1
+      if (StepAndGetOpcode() != AArch64::EORXrs)
+        return std::nullopt;
+      TestedReg = It->getOperand(1).getReg();
+      if (It->getOperand(0).getReg() != ScratchReg ||
+          It->getOperand(2).getReg() != TestedReg ||
+          It->getOperand(3).getImm() != 1)
+        return std::nullopt;
+
+      return std::make_pair(TestedReg, &*It);
+    }
+  }
+
+  MCPhysReg getAuthCheckedReg(const MCInst &Inst,
+                              bool MayOverwrite) const override {
+    // Cannot trivially reuse AArch64InstrInfo::getMemOperandWithOffsetWidth()
+    // method as it accepts an instance of MachineInstr, not MCInst.
+    const MCInstrDesc &Desc = Info->get(Inst.getOpcode());
+
+    // If signing oracles are considered, the particular value left in the base
+    // register after this instruction is important. This function checks that
+    // if the base register was overwritten, it is due to address write-back.
+    //
+    // Note that this function is not needed for authentication oracles, as the
+    // particular value left in the register after a successful memory access
+    // is not important.
+    auto ClobbersBaseRegExceptWriteback = [&](unsigned BaseRegUseIndex) {
+      MCPhysReg BaseReg = Inst.getOperand(BaseRegUseIndex).getReg();
+      unsigned WrittenBackDefIndex = Desc.getOperandConstraint(
+          BaseRegUseIndex, MCOI::OperandConstraint::TIED_TO);
```

Comment:
This feels like maybe the binary analysis tool here is looking into too much detail of instruction descriptions. I'm wondering if it would be better to have one or a few functions added to `AArch64InstrInfo`, so that special knowledge about "parsing"/"interpreting" machineOperands might be more all in the same place?
I think I've tried to adapt tablegen to have a bit more of this kind of information in my prototype for the stack clash scanner directly generated from tablegen files. But that is a really big change, and may not cover what you need here exactly. A first step could be to define the functions/methods/API in `AArch64InstrInfo` you need to implement this function so that you don't have to query e.g. `getOperandConstraint` etc here?

Reply:
I didn't have a specific concrete design in mind.
Initially, I'm wondering if a method like `getBaseRegister` or something similar could be added to `AArch64InstrInfo` that returns the base register (or `NoRegister` when the instruction does not have a base register) when given an MCInst or MachineOperand or similar.
I'm not sure if that makes sense?
I hope it would simplify the code here, and might encourage reuse of the functionality in case it's needed in other parts of the compiler.
If later on new AArch64 instructions are added for which the base register needs to be computed in a different way, there's a higher chance `AArch64InstrInfo` will get adapted than this code deep inside BOLT.

Reply:
Oh yes, it probably would be a whole separate PR, that would need to be reviewed by an expert in that area.
Maybe the best thing to progress would be to just add a FIXME here and fix the FIXME in a separate PR?

Reply:
Added a FIXME in f5455c6b4b0d0ae07dd8062643e32bfba4111946.